From e6dc8c220bfb28dd98cedc379ea9e1e5525e72f3 Mon Sep 17 00:00:00 2001
From: Andy Grover <agrover@redhat.com>
Date: Fri, 5 Sep 2014 16:36:28 -0400
Subject: [target] iscsi: Avoid rejecting incorrect ITT for Data-Out

Message-id: <1409934989-29941-31-git-send-email-agrover@redhat.com>
Patchwork-id: 91484
O-Subject: [RHEL7.1 PATCH e-stor 30/31] iscsi-target: Avoid rejecting incorrect ITT for Data-Out
Bugzilla: 1129387
RH-Acked-by: Chris Leech <cleech@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

From: Nicholas Bellinger <nab@linux-iscsi.org>

commit 97c99b47ac58bacb7c09e1f47d5d184434f6b06a upstream.

This patch changes iscsit_check_dataout_hdr() to dump the incoming
Data-Out payload when the received ITT is not associated with a
WRITE, instead of calling iscsit_reject_cmd() for the non WRITE
ITT descriptor.

This addresses a bug where an initiator sending an Data-Out for
an ITT associated with a READ would end up generating a reject
for the READ, eventually resulting in list corruption.

Reported-by: Santosh Kulkarni <santosh.kulkarni@calsoftinc.com>
Reported-by: Arshad Hussain <arshad.hussain@calsoftinc.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/target/iscsi/iscsi_target.c b/drivers/target/iscsi/iscsi_target.c
index 2173451..b1bdd4e 100644
--- a/drivers/target/iscsi/iscsi_target.c
+++ b/drivers/target/iscsi/iscsi_target.c
@@ -1313,7 +1313,7 @@ iscsit_check_dataout_hdr(struct iscsi_conn *conn, unsigned char *buf,
  if (cmd->data_direction != DMA_TO_DEVICE) {
   pr_err("Command ITT: 0x%08x received DataOUT for a"
    " NON-WRITE command.\n", cmd->init_task_tag);
-  return iscsit_reject_cmd(cmd, ISCSI_REASON_PROTOCOL_ERROR, buf);
+  return iscsit_dump_data_payload(conn, payload_length, 1);
  }
  se_cmd = &cmd->se_cmd;
  iscsit_mod_dataout_timer(cmd);
-- 
1.7.1